From cf6b662b8fe450ea7dc6eb5adb2b878971ce0d67 Mon Sep 17 00:00:00 2001
From: George Joseph <gjoseph@digium.com>
Date: Wed, 10 Jan 2018 08:48:32 -0700
Subject: [PATCH] sip_msg: Prevent crash on header without vptr

Occasionally a header with no vptr gets into the list of header.  This
causes a crash when printing them.  We still need to figure out how
the header got there but this patch at least prevents the crash by checking
for a non-NULL vptr before attempting to call its print function.
---
 pjsip/src/pjsip/sip_msg.c | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/pjsip/src/pjsip/sip_msg.c b/pjsip/src/pjsip/sip_msg.c
index 76d07b563..194b5ee01 100644
--- a/pjsip/src/pjsip/sip_msg.c
+++ b/pjsip/src/pjsip/sip_msg.c
@@ -26,6 +26,7 @@
 #include <pj/string.h>
 #include <pj/pool.h>
 #include <pj/assert.h>
+#include <pj/log.h>
 #include <pjlib-util/string.h>
 
 PJ_DEF_DATA(const pjsip_method) pjsip_invite_method =
@@ -471,6 +472,8 @@ PJ_DEF(pj_ssize_t) pjsip_msg_print( const pjsip_msg *msg,
 
 	    *p++ = '\r';
 	    *p++ = '\n';
+	} else {
+	    PJ_LOG(1, ("sip_msg", "Header with no vptr encountered!!  Current buffer: %.*s", (int)(p-buf), buf));
 	}
     }
 
@@ -578,6 +581,9 @@ PJ_DEF(void*) pjsip_hdr_shallow_clone( pj_pool_t *pool, const void *hdr_ptr )
 PJ_DEF(int) pjsip_hdr_print_on( void *hdr_ptr, char *buf, pj_size_t len)
 {
     pjsip_hdr *hdr = (pjsip_hdr*) hdr_ptr;
+    if (!hdr->vptr) {
+        return 0;
+    }
     return (*hdr->vptr->print_on)(hdr_ptr, buf, len);
 }
 
-- 
2.14.3

